theano/sandbox/cuda/tests/test_dnn.py

```diff
@@ -450,43 +452,132 @@ def test_pooling_opt():
                 for n in f.maker.fgraph.toposort()])
 
 
-def test_log_softmax():
-    # This is a test for an optimization that depends on CuDNN v3 or
-    # more recent. Don't test if the CuDNN version is too old.
-    if not cuda.dnn.dnn_available() or cuda.dnn.version() < (3000, 3000):
-        raise SkipTest(cuda.dnn.dnn_available.msg)
+class test_DnnSoftMax(test_nnet.test_SoftMax):
+    gpu_op = dnn.GpuDnnSoftmax
+    gpu_grad_op = dnn.GpuDnnSoftmaxGrad
+    mode = mode_with_gpu
+    do_0 = False
+    topo_idx = -3
 
-    x = T.ftensor4()
-    softmax_out = dnn.GpuDnnSoftmax('bc01', 'accurate', 'channel')(x)
-    log_out = T.log(T.as_tensor_variable(softmax_out))
-
-    f = theano.function([x], log_out, mode=mode_with_gpu)
-
-    # Ensure that the optimization has been applied
-    dnn_softmax_nodes = [n for n in f.maker.fgraph.toposort() if
-                         isinstance(n.op, cuda.dnn.GpuDnnSoftmax)]
-    assert len(dnn_softmax_nodes) == 1
-    assert dnn_softmax_nodes[0].op.algo == "log"
-
-    # Ensure that the output of the function is valid
-    input_shapes = [(3, 4, 5, 6),
-                    (1025, 2, 3, 4),
-                    (2, 1025, 3, 4),
-                    (2, 3, 1025, 4),
-                    (2, 3, 4, 1025),
-                    (66000, 2, 3, 4),
-                    (2, 66000, 3, 4),
-                    (2, 3, 66000, 4),
-                    (2, 3, 4, 66000)]
-
-    for inp_shape in input_shapes:
-        input_val = numpy.random.normal(0, 1, inp_shape).astype("float32")
-
-        out = f(input_val)
-        expected_out = numpy.log(numpy.exp(input_val) /
-                                 numpy.exp(input_val).sum(1)[:, None, :, :])
-
-        utt.assert_allclose(out, expected_out)
+    def setUp(self):
+        if not cuda.dnn.dnn_available():
+            raise SkipTest(cuda.dnn.dnn_available.msg)
+        utt.seed_rng()
+
+    def test_dnn_softmax_grad(self):
+        softmax_op = dnn.GpuDnnSoftmax('bc01', 'accurate', 'channel')
+
+        x_val = numpy.random.normal(0, 1, (3, 4, 2, 5)).astype('float32')
+        x_val2 = numpy.random.normal(0, 1, (3, 4, 1, 1)).astype('float32')
+
+        utt.verify_grad(softmax_op, [x_val])
+
+        # Gradient is broken for (n, c, 1, 1) in v3 rc1
+        if cuda.dnn.version() == (2000, 2000):
```

Comment:
I would do 

... != (3000, 3000), next version should work, so we don't need to update.